bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/framework/internal/reliablefile/ReliableFile.java

```diff
@@ -681,6 +696,39 @@ public static void cleanupGenerations(File base) {
 		}
 	}
 
+	/*
+	 * Implementation note: This implementation differs from File.createTempFile by
+	 * avoiding usage of SecureRandom to generate unique file names.
+	 * 
+	 * Any usage of this must be used in a context that is not sensitive to outside
+	 * guessing of the temporary file name.
+	 */
+	public static File createTempFile(String prefix, String suffix, File directory) throws IOException {
+		if (directory == null) {
+			throw new IllegalArgumentException("No directory specified."); //$NON-NLS-1$
+		}
+		if (prefix == null) {
+			throw new IllegalArgumentException("No prefix specified."); //$NON-NLS-1$
+		}
+		if (suffix == null) {
+			suffix = ".tmp"; //$NON-NLS-1$
+		}
+		File f;
+		do {
+			// just increments a global integer between 1 and 100000
+			f = new File(directory, prefix + nextTemp.getAndUpdate(n -> {
```

Comment:
why not simply incrementAndGet ? Is there really a chance of overflowing here?

Reply:
This is a global static that would be used by any number of frameworks running in the same JVM.  We have certainly run into overflow on other areas of the framework.  I'm not willing to take the risk here.

Reply:
If overflowing is a potential problem AtomicLong could be used instead. If that overflows the drive does it as well.

Or is it intended to limit the number of existing temp-files? If yes, maybe `File.deleteOnExit()` could be an option?

Reply:
Keep in mind this is used for staging quick file operations.  So something like this:

1. create temp file
2. write full content of file
3. move temp file to final destination

So these are not really "temp" files in that regard.  The are staging files where we write content.  Then we make sure all the content is written before we can safely place it to its final destination.  It is intended to protect us from dead files if there is something like a hard crash or power outage in the middle of writing.  The idea is the final move operation can be "atomic" and happens after we know all has been successfully written.

I wanted to only use positive numbers in the file name and really want to avoid overflow.  We "can" use Long which likely would delay any overflow to eternity, but seems like it could start making the staging file names needlessly long over time.  I don't see the reason to risk it here.  Especially because this can be used by multiple framework instances at the same time.  Is the objection that the overflow logic is to complex or incorrect?

Reply:
The other thing to mention is that the usage in the framework is not to keep many of these staging files in existence over time.  So it is not like the file system is going to run out of inodes before we overflow.

Reply:
Thank you for clarifying the use-case. So too many existing files is likely not a problem (unless one accumulates many hard failures/power-losses of the framework over time).


> I wanted to only use positive numbers in the file name and really want to avoid overflow. We "can" use Long which likely would delay any overflow to eternity, but seems like it could start making the staging file names needlessly long over time. I don't see the reason to risk it here. Especially because this can be used by multiple framework instances at the same time. 

`Long.MAX_VALUE` is `9223372036854775807`, `Integer.MAX_VALUE` is `2147483647`. Long is clearly longer but I would say it is acceptable (at least on Linux where you have no max-path length) since it is unlikely high numbers are reached quickly.

>Is the objection that the overflow logic is to complex or incorrect?

I cannot speak for Christoph, but I would say that less/simpler code is better, if it is sufficient. But the logic look correct.
In the end I cannot clearly tell what's right or wrong here.

Reply:
The more I thought about this the more I think we should fail with an IOException if an excessive amount of retries are done.  Clearly something is wrong if we looped over 100000 file names and could not find one that allowed us to create a new file in its location.  Probably could be even much less than 100000 file names, even 100 existing files seems way to much.  But we can start there and adjust down.